WIP - (MODULES-4640) - Generate metadata POT file and Rake Task

pmcmaw opened the pull request:
Using Ruby to generate a POT file for metadata.json. This also creates a rake task to generate the POT file. This will allow us to automate keeping the metadata.json POT file up to date.

This has been tested on tagmail and accounts.

eputnam commented on lib/tasks/gettext.rake:
```diff
@@ -83,6 +145,17 @@ namespace :gettext do
+
+  desc 'Merge POT Files'
+  task :merge_pot_files do
+    system("msgmerge -U #{metadata_pot_file_path} #{pot_file_path} --output-file=locales/output.pot")
```
the name of the new POT needs to be changed back to `GettextSetup.config['project_name']` at some point. Perhaps just another line or two in this task?

pmcmaw replied:
Apologies Eric, this merge POT files was something I was playing with to try out and have forgotten to take it out. My bad. Sorting this now.

eputnam commented:
What do you think about appending the description and summary lines into an existing POT instead of creating an entirely new POT file and merging it in?

pmcmaw commented:
@eputnam, 
My initial thought was eventually there will be 3 different processes happening eg:
1) Create metadata POT file
2) Create Ruby POT file
3) Create Puppet POT file
That it may be easier to keep all of the information separate. It will also avoid running into issues when something has been updated. Eg the summary and description were appended to the already existing POT file, if the summary has been updated, it may be more difficult to update the POT file, whereas if it is in its own file, if an update has occurred it is easy just to regenerate the files then merge. Although definitely open to change if you disagree or have a different opinion.

tphoney commented:
I would vote for merging the sub-pot files. As there may be future processes that will be adding to the master-pot file. That would mean they would each have to write their own custom merge. 
Using a single merge process of constituent sub-pot files is a more consistent approach. 

Code looks good 👍

HelenCampbell commented:
I don't think there are massive benefits to either appending or generating in this case. However that would make me lean towards TP's consistency point. In my head it would be cleaner to rewrite the POT than to try and update current fields/ handle merge or deletion, however I've had little technical experience around something like that, willing to be proven wrong.
So overall I'm for the sub-pots and merge - and happy with code.

smcclellan commented on .rubocop.yml:
```diff
@@ -26,7 +26,8 @@ Metrics/BlockLength:
 Metrics/ModuleLength:
   Enabled: false
-
+Style/FileName:
```
This change isn't needed if you rebase this PR.

smcclellan commented on lib/tasks/gettext.rake:
```diff
@@ -52,6 +52,68 @@ namespace :gettext do
            files_to_translate.join(' '))
   end
 
+  def metadata_pot_file_path
```
Can we move all this code into `lib/`? That way we can write some tests against the generated file.

smcclellan replied:
Rather, move it into an `.rb` file for ease of testing. The other PR in this repo creates a `task_helper.rb` file, which seems a sensible way about it.

smcclellan commented on lib/tasks/gettext.rake:
```diff
@@ -52,6 +52,68 @@ namespace :gettext do
+
+  def generate_new_pot_metadata
+    open(metadata_pot_file_path, 'w') do |f|
+      f << "#\n"
```
Can we just use a HEREDOC for all of these? For example:

```
f << <<-DOC
#
#, fuzzy
# msgid ""
# msgstr ""
"Project-Id-Version: PACKAGE VERSION\n"
...
DOC
```

That way it's more like editing an actual file whenever we change this.

smcclellan commented on lib/tasks/gettext.rake:
```diff
@@ -52,6 +52,68 @@ namespace :gettext do
+    summary_metadata['summary'] if summary_metadata.key?('summary')
+  end
+
+  def current_time
```
This helper method is only used once, probably can be called in-line.

smcclellan commented on lib/tasks/gettext.rake:
```diff
@@ -52,6 +52,68 @@ namespace :gettext do
+  end
+
+  def load_metadata_information
+    metadata_file = 'metadata.json'
```
To avoid this being called so many times, opening the file each time, I suggest doing something like this:

```
@metadata ||= begin
  metadata_file = 'metadata.json'
  file = open(metadata_file)
  json = file.read
  JSON.parse(json)
end
```

smcclellan replied:
If this method only gets called once though (after the other fixes), this matters less.

smcclellan commented on lib/tasks/gettext.rake:
```diff
@@ -52,6 +52,68 @@ namespace :gettext do
+    end
+  end
+
+  def load_metadata_information
```
If the below comment is addressed, this can just be called `metadata` rather than the action of loading.

smcclellan commented on lib/tasks/gettext.rake:
```diff
@@ -52,6 +52,68 @@ namespace :gettext do
+      if metadata.key?('summary')
+        f << "#. metadata.json\n"
+        f << "#: .summary\n"
+        f << "msgid \"#{metadata_summary}\"\n"
```
This should probably just be metadata['summary'].

smcclellan commented on lib/tasks/gettext.rake:
```diff
@@ -52,6 +52,68 @@ namespace :gettext do
+    JSON.parse(json)
+  end
+
+  def metadata_description
```
This and the below method shouldn't be needed.

pmcmaw commented:
@smcclellan I have now addressed your feedback, it was much appreciated, therefore thank you for taking the time. 

I have had a struggle with rubocop, and I didn't want to make changes in the gemspec therefore I disabled the cop for it. If you would prefer the changed to be made I can run rubocop with the -a flag to auto correct them, I just didn't want to go ahead and make changes without consulting you.

smcclellan commented on lib/metadata_pot/metadata_pot.rb:
```diff
@@ -0,0 +1,55 @@
+"Content-Transfer-Encoding: 8bit\\n"
+"X-Generator: Translate Toolkit 2.0.0\\n"
+#
+      DOC
```
Minor nit (which may be what rubocop was complaining about): I think this `DOC` should line up with `<<-DOC`.

pmcmaw replied:
So the cop is wanting me to indent all of the content going into the file by 2 spaces (which I finally realized). When I do this, the file contents is then written with an indent of 2 spaces. I am able to over come this by using <<~DOC although this functionality has only been implemented in Ruby 2.3.

smcclellan replied:
That's fine. I'll merge now and do some refactoring as a separate commit. In all, this functionality looks fine to me; I just want to refactor so that it's easier to add tests, and that changes can be quicker in the future.

smcclellan commented:
Refactoring PR is here: https://github.com/puppetlabs/gettext-setup-gem/pull/36
